bound/kt/src/main/kotlin/tbdex/sdk/messages/Close.kt

```diff
@@ -8,50 +8,51 @@ import tbdex.sdk.rust.CloseDataData as RustCoreCloseData
 
 typealias CloseData = RustCoreCloseData
 
-class Close: Message, ReplyToMessage {
+class Close private constructor(
+    val metadata: MessageMetadata,
+    val data: CloseData,
+    val signature: String,
+    internal val rustCoreClose: RustCoreClose
+): Message, ReplyToMessage {
     init {
         SystemArchitecture.set() // ensure the sys arch is set for first-time loading
     }
 
-    val metadata: MessageMetadata
-    val data: CloseData
-    val signature: String
-
-    val rustCoreClose: RustCoreClose
-
-    constructor(
-        bearerDid: BearerDid,
-        to: String,
-        from: String,
-        exchangeId: String,
-        data: CloseData,
-        protocol: String,
-        externalId: String? = null
-    ) {
-        this.rustCoreClose = RustCoreClose(bearerDid.rustCoreBearerDid, to, from, exchangeId, data, protocol, externalId)
-
-        this.metadata = rustCoreClose.getData().metadata
-        this.data = rustCoreClose.getData().data
-        this.signature = rustCoreClose.getData().signature
+    companion object {
+        fun create(
+            to: String,
+            from: String,
+            exchangeId: String,
+            data: CloseData,
+            protocol: String? = null,
+            externalId: String? = null
+        ): Close {
+            val rustCoreClose = RustCoreClose.create(to, from, exchangeId, data, protocol, externalId)
+            val rustCoreData = rustCoreClose.getData()
+            return Close(rustCoreData.metadata, rustCoreData.data, rustCoreData.signature, rustCoreClose)
+        }
+
+        fun fromJsonString(json: String): Close {
+            val rustCoreClose = RustCoreClose.fromJsonString(json)
+            val rustCoreData = rustCoreClose.getData()
+            return Close(rustCoreData.metadata, rustCoreData.data, rustCoreData.signature, rustCoreClose)
+        }
+
+        internal fun fromRustCoreClose(rustCoreClose: RustCoreClose): Close {
+            val rustCoreData = rustCoreClose.getData()
+            return Close(rustCoreData.metadata, rustCoreData.data, rustCoreData.signature, rustCoreClose)
+        }
     }
 
-    constructor(json: String) {
-        this.rustCoreClose = RustCoreClose.fromJsonString(json)
-
-        this.metadata = rustCoreClose.getData().metadata
-        this.data = rustCoreClose.getData().data
-        this.signature = rustCoreClose.getData().signature
+    fun toJsonString(): String {
+        return this.rustCoreClose.toJsonString()
     }
 
-    constructor(rustCoreClose: RustCoreClose) {
-        this.rustCoreClose = rustCoreClose
-
-        this.metadata = rustCoreClose.getData().metadata
-        this.data = rustCoreClose.getData().data
-        this.signature = rustCoreClose.getData().signature
+    fun sign(bearerDid: BearerDid) {
+        this.rustCoreClose.sign(bearerDid.rustCoreBearerDid)
     }
 
-    fun toJson(): String {
-        return this.rustCoreClose.toJson()
+    fun verify() {
+        this.rustCoreClose.verify()
```

Comment:
did we try throwing an error and it reads nicely ?

Reply:
we should prioritize friendly errors next actually!